### PR TITLE
Refactored EfCoreMigrationsDialog

### PR DESCRIPTION
**Pull request by Herdo**
Refactored to MVVM
- Currently missing UnitTests
- Needs a lot of manual testing due to complexity
- Needs some more refactoring after functionality is OK

### CODE REVIEW

**Comment by Herdo** [comment]
@ErikEJ That was quiet the complex dialog. It'll need some more changes.
Would be great if you could provide example projects to debug all the functionalities, or we could do a one-on-one session to finalize the dialog 😄 
Maybe this could leave the "preview" status afterwards 😉

**Comment by ErikEJ** [comment]
Thanks for this! I will surely leave this feature in "preview" state. I will get your branch and try it out.

File: src/GUI/EFCorePowerTools/EFCorePowerTools.csproj [hunk 1]
```diff
@@ -293,6 +296,7 @@
       <HintPath>..\lib\Microsoft.Data.Tools.Utilities.dll</HintPath>
       <Private>False</Private>
     </Reference>
+    <Reference Include="Microsoft.EntityFrameworkCore.Design, Version=2.2.1.0, Culture=neutral, PublicKeyToken=adb9793829ddae60, processorArchitecture=MSIL" />
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/EFCorePowerTools.csproj]: Is this really needed? If so, use 2.2.0

Comment (Herdo) [reply src/GUI/EFCorePowerTools/EFCorePowerTools.csproj]: Hm, that might have slipped in there thanks to ReSharper. I'll remove it.

**Comment by ErikEJ** [comment]
You should be able to use the hwo-to here to test functionality: https://github.com/ErikEJ/EFCorePowerTools/wiki/Migrations-Tool

**Comment by Herdo** [comment]
@ErikEJ I'm still struggling to get this running. Neither the currently released version (2.2.12.0) nor the PR version seem to work. They both throw the same error:
grafik

Here's the demo project:
Any idea what it's missing?
[ConsoleApp3.zip](https://github.com/ErikEJ/EFCorePowerTools/files/3035350/ConsoleApp3.zip)

**Comment by ErikEJ** [comment]
Did you follow the steps on the wiki post?

**Comment by ErikEJ** [comment]
Refresh the PR with latest master, or use EF Core 2.2.2 iso 2.2.3

**Comment by Herdo** [comment]
I did follow the example linked, until (excluding) the step(s) of "Create your database".
grafik

As I understand, the dialog will do that part of work, even for the initial creation?
Or do I need to add the initial creation, therefore including the step(s) of "Create your database"?

I'll refresh the PR with the latest master...

**Comment by ErikEJ** [comment]
Yes, migrations will also create the database for you.

**Comment by Herdo** [comment]
Already using EF Core 2.2.3.

**Comment by ErikEJ** [comment]
>> Already using EF Core 2.2.3.

Not sure what you mean? But Power Tools must be explicitly updated to support 2.2.3

**Comment by Herdo** [comment]
> 
> 
> > > Already using EF Core 2.2.3.
> 
> Not sure what you mean? But Power Tools must be explicitly updated to support 2.2.3

Oh. So the example solution must reference 2.2.2? Let me give it a try...

**Comment by Herdo** [comment]
Changing the reference to 2.2.0 helps.
Even though it seems that the rebase messed up a little bit...

**Comment by Herdo** [comment]
@ErikEJ Getting the following exception while adding a new migration, both on _ErikEJ/master_ as well as on _Herdo/dialogs_ :

> Error:
System.InvalidOperationException: No service for type 'Microsoft.EntityFrameworkCore.Migrations.Design.MigrationsScaffolder' has been registered.
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at ReverseEngineer20.EfCoreMigrationsBuilder.AddMigration(String name, String outputPath, String projectPath, String nameSpace, DbContext context) in C:\Code\EFCorePowerTools\src\GUI\efpt\EFCoreMigrationsBuilder.cs:line 169
   at ReverseEngineer20.EfCoreMigrationsBuilder.AddMigration(DbContext context, String outputPath, String projectPath, String name, String nameSpace) in C:\Code\EFCorePowerTools\src\GUI\efpt\EFCoreMigrationsBuilder.cs:line 144
   at ReverseEngineer20.EfCoreMigrationsBuilder.BuildMigrationResult(String outputPath, String projectPath, String contextName, Boolean addMigration, Boolean scriptMigration, String migrationIdentifier, String nameSpace) in C:\Code\EFCorePowerTools\src\GUI\efpt\EFCoreMigrationsBuilder.cs:line 57
   at ReverseEngineer20.Program.Main(String[] args) in C:\Code\EFCorePowerTools\src\GUI\efpt\Program.cs:line 38

Still using the example project attached yesterday, with the EF Core reference changed to 2.2.0.

**Comment by ErikEJ** [comment]
I Will give it a try with current master...

**Comment by ErikEJ** [comment]
Apparently no one uses this feature! :smile: 
I fixed the bug in the latest master, and also modified the BuildModelResult code:

    result.Add(parts[0].Trim(),  parts.Length > 1 ? parts[1].Trim() : string.Empty);

**Comment by ErikEJ** [comment]
https://github.com/ErikEJ/EFCorePowerTools/commit/151687ea3804037ea59d27bde9ae2dbcc1892871

**Comment by ErikEJ** [comment]
@Herdo FYI: https://github.com/ErikEJ/EFCorePowerTools/pull/203 (sorry)

File: src/GUI/efpt21.core/efpt21.core.csproj [hunk 2]
```diff
@@ -22,8 +22,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="2.1.0" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.1.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="2.1.8" />
```
Comment (ErikEJ) [inline src/GUI/efpt21.core/efpt21.core.csproj]: Revert to 2.1.0

File: src/GUI/efpt21.core/efpt21.core.csproj [hunk 3]
```diff
@@ -22,8 +22,8 @@
...
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.1.8" />
   </ItemGroup>
```
Comment (ErikEJ) [inline src/GUI/efpt21.core/efpt21.core.csproj]: Revert to 2.1.0

File: src/GUI/efpt22.core/efpt22.core.csproj [hunk 4]
```diff
@@ -22,8 +22,8 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="2.2.0" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Design" Version="2.2.4" />
```
Comment (ErikEJ) [inline src/GUI/efpt22.core/efpt22.core.csproj]: Revert to 2.2.0

File: src/GUI/efpt22.core/efpt22.core.csproj [hunk 5]
```diff
@@ -22,8 +22,8 @@
...
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="2.2.4" />
```
Comment (ErikEJ) [inline src/GUI/efpt22.core/efpt22.core.csproj]: Revert to 2.2.0

**Comment by ErikEJ** [comment]
@Herdo Are you able to do testing now?

**Comment by Herdo** [comment]
@ErikEJ Testing works now 👍

**Comment by ErikEJ** [comment]
Sounds great - ready to merge??

**Comment by Herdo** [comment]
@ErikEJ Have a look 😉 

New initial view:
grafik

Upate database view:
grafik

New in-sync view:
grafik

**Comment by Herdo** [comment]
I hope the margin on the unicorn picture is not a problem for you. I can change that, though, if you want.

**Comment by ErikEJ** [comment]
LGTM !! Let's merge it in then - thanks for your patience

**Pull request merged**